apps/full-site-editing/full-site-editing-plugin/wpcom-block-editor-nav-sidebar/src/components/nav-sidebar/index.tsx

```diff
@@ -63,6 +62,57 @@ function WpcomBlockEditorNavSidebar() {
 		prevIsOpen.current = isOpen;
 	}, [ isOpen, prevIsOpen, setSidebarClosing ] );
 
+	const [ isListScrolling, setIsListScrolling ] = useState( false );
+
+	const observer = useRef< IntersectionObserver >();
+	const itemRefs = useRef< ( HTMLElement | null )[] >( [] );
+
+	const containerMount = ( el: HTMLDivElement | null ) => {
+		if ( el ) {
+			el.focus();
+		}
+
+		if ( observer.current ) {
+			observer.current.disconnect();
+		}
+
+		if ( ! window.IntersectionObserver ) {
+			return;
+		}
+
+		observer.current = new window.IntersectionObserver(
+			( entries ) => {
+				// If every iten is currently visible, then we're not scrolling
+				const notScrolling = entries.every( ( entry ) => entry.isIntersecting );
+				setIsListScrolling( ! notScrolling );
+			},
+			{
+				root: el,
+				threshold: [ 1 ], // We want to fire the above callback when an entry isn't 100% visible
+			}
+		);
+	};
+
+	const itemMount = ( el: HTMLElement | null, itemIndex: number ) => {
+		itemRefs.current[ itemIndex ] = el;
+	};
+
+	useEffect( () => {
+		if ( ! observer.current ) {
+			return;
+		}
+
+		observer.current.disconnect();
+
+		const nonSparse = itemRefs.current.filter( Boolean ) as HTMLElement[];
+
+		// We only need to observe the first and last item in the list, might be better for performance
+		const firstItem = nonSparse[ 0 ];
+		const lastItem = nonSparse[ nonSparse.length - 1 ];
+		if ( firstItem ) observer.current.observe( firstItem );
+		if ( lastItem ) observer.current.observe( lastItem );
+	}, [ items ] );
```

Comment:
hmmm I think observer and itemRefs should be dependencies here

I was trying to work out why observer uses `useRef` rather than being just a plain variable. I think the benefit of useRef is that we can make observer a dependency of this `useEffect` so that we can be guaranteed that useRef will be invoked after the observer is initialized.

Reply:
The rationale for using a ref for the observer is that if it was a local variable it'd be a different variable each time the component function is called. The ref will persist for the lifetime of the component, kinda like a member variable in a class component.

You're right about it being a dependency though, I'm surprised I'm not getting a lint error.

Reply:
ahh yes that makes sense about useRef, thanks. 👍